File: AngularRequireSignalR/App/Chat/Models/ChatRoom.js
define(function (require) {

    var Message = require('Chat/Models/Message');

    var ChatRoom = function() {
        this._userLookup = {};
        this.users = [];
        this.messages = [];
    };

    ChatRoom.prototype.addUser = function(user) {
        this.users.push(user);
        this._userLookup[user.id] = user;
    };
    ChatRoom.prototype.addMessage = function (m) {

        var message = new Message(this._userLookup[m.UserId], m.Message);
        this.messages.push(message);
        return message;
    };

    return ChatRoom;
});

File: AngularRequireSignalR/Hubs/ViewModels/ChatRoom.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AngularRequireSignalR.Hubs.ViewModels
{
    public interface IChatRoom
    {
        List<ChatUser> Users { get; }
        List<ChatMessage> Messages { get; }
        ChatUser AddUser(string userId, string userName);
        ChatMessage AddMessage(string userId, string text);
    }

    public class ChatRoom : IChatRoom
    {
        private readonly List<ChatUser> _users = new List<ChatUser>();
        private readonly List<ChatMessage> _messages = new List<ChatMessage>();

        public List<ChatUser> Users
        {
            get { return _users; }
        }

        public List<ChatMessage> Messages
        {
            get { return _messages; }
        }

        public ChatUser AddUser(string userId, string userName)
        {
            var user = new ChatUser { Id = userId ?? Guid.NewGuid().ToString(), Name = userName };
            Users.Add(user);
            return user;
        }

        public ChatMessage AddMessage(string userId, string text)
        {
            var message = new ChatMessage { UserId = userId, Message = text, TimeStamp = DateTime.Now };
            Messages.Add(message);
            return message;
        }
    }
}